securedrop/journalist_app/api2/types.py

```diff
@@ -0,0 +1,97 @@
+from dataclasses import dataclass, field
+from enum import IntEnum, StrEnum, auto
+from typing import (
+    Any,
+    Dict,
+    List,
+    NewType,
+    Set,
+    Union,
+)
+
+Version = NewType("Version", str)
+
+
+# TODO: generic UUID[T] in Python 3.12
+SourceUUID = NewType("SourceUUID", str)
+ItemUUID = NewType("ItemUUID", str)
+JournalistUUID = NewType("JournalistUUID", str)
+
+
+EventID = NewType("EventID", str)  # int, but opaque on the wire
+
+
+class EventType(StrEnum):
+    REPLY_SENT = auto()
+
+
+class EventStatusCode(IntEnum):
+    OK = 200
+    BadRequest = 400
+    NotImplemented = 501
+
+
+@dataclass
+class Index:
+    # Source metadata, optionally filtered by `source_prefix`:
+    sources: Dict[SourceUUID, Version] = field(default_factory=dict)
+    items: Dict[ItemUUID, Version] = field(default_factory=dict)
+
+    # Non-source metadata (always returned):
+    journalists: Dict[JournalistUUID, Version] = field(default_factory=dict)
+
+
+@dataclass
+class SourceTarget:
+    source_uuid: SourceUUID
+    version: Version
+
+
+@dataclass
+class ItemTarget:
+    item_uuid: ItemUUID
+    version: Version
+
+
+@dataclass
+class Event:
+    id: EventID
+    target: Union[SourceTarget, ItemTarget]
+    type: EventType
+    data: Any = field(default_factory=dict)
+
+
+@dataclass
+class EventResult:
+    event_id: EventID
+    status: EventStatusCode
+
+    # Changed:
+    sources: Dict[SourceUUID, Any] = field(default_factory=dict)
+    items: Dict[ItemUUID, Any] = field(default_factory=dict)
+
+
+@dataclass
+class BatchRequest:
+    # Source metadata:
+    sources: Set[SourceUUID] = field(default_factory=set)
+    items: Set[ItemUUID] = field(default_factory=set)
+
+    # Non-source metadata:
+    journalists: Set[JournalistUUID] = field(default_factory=set)
+
+    # Events submitted by the client:
+    events: List[Event] = field(default_factory=list)
+
+
+@dataclass
+class BatchResponse:
```

Comment:
How will we represent the absence of data, i.e. a source being deleted?

Reply:
Do we need to?  If the client has submitted an event like `source_deleted(event_id=123456, source_uuid="abcdef")`, then it's already carried it out the deletion client-side; it just needs to know the server accepted it:

1. ```{"events": {"id": 123456, "status": 200}, ...}``` → event `123456` accepted

If the server has deleted source `abcdef`, then in the next sync the client will receive an index missing source `abcdef` and know to delete it, per freedomofpress/securedrop-client#2716.

Now, we could also return here:

2. ```{..., "sources": {"abcdef": null}}``` → delete source `abcdef`

(1) should be sufficient, but we can do (2) as well if we want to be equally explicit about the effects of all events.